app/src/main/java/tech/relaycorp/gateway/domain/endpoint/CollectParcels.kt

```diff
@@ -0,0 +1,76 @@
+package tech.relaycorp.gateway.domain.endpoint
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.onEach
+import tech.relaycorp.gateway.common.Logging.logger
+import tech.relaycorp.gateway.data.database.StoredParcelDao
+import tech.relaycorp.gateway.data.disk.DiskMessageOperations
+import tech.relaycorp.gateway.data.disk.MessageDataNotFoundException
+import tech.relaycorp.gateway.data.model.MessageAddress
+import tech.relaycorp.gateway.data.model.RecipientLocation
+import tech.relaycorp.gateway.data.model.StoredParcel
+import tech.relaycorp.gateway.domain.DeleteParcel
+import java.io.InputStream
+import java.util.UUID
+import java.util.logging.Level
+import javax.inject.Inject
+
+/*
+    This class keeps the ParcelCollection state for one connection.
+    It should not be re-used between ParcelCollection connections.
+ */
+class CollectParcels
+@Inject constructor(
+    private val storedParcelDao: StoredParcelDao,
+    private val diskMessageOperations: DiskMessageOperations,
+    private val deleteParcel: DeleteParcel
+) {
+
+    private val parcelsSent = mutableListOf<StoredParcel>()
+    private val parcelsWaitingAck = mutableMapOf<String, StoredParcel>()
+
+    private val _noParcelsLeft = MutableStateFlow(false)
+    val noParcelsToDeliverOrAck: Flow<Boolean> = _noParcelsLeft
```

Comment:
That part is "magical" :P

It's not clear from the logic, because it happens indirectly, through the reactive DB query. When you receive an ACK, the stored parcel gets deleted, so the `storedParcelDao.listForRecipients` returns again.